### ISSUE TITLE
Building Against Openssl 1.1.0h Breaks in keyraw.c

### ISSUE THREAD

**Issue by iamperson347** [getdnsapi/getdns#401]
Hi all,

Openwrt appears to be gearing up to move to openssl 1.1*. I built openssl 1.1.0h using the proposed patch and attempted to build getdns against that with the following error (snippet below). 

Please let me know if you need additional details and I will try to dig them up.

../libtool --quiet --tag=CC --mode=compile ccache_cc -I. -I. -I./util/auxiliary -I./../stubby/src -Os -pipe -mno-branch-likely -mips32r2 -mtune=24kc -fno-caller-saves -fno-plt -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -msoft-float -mips16 -minterlink-mips16 -iremap/working/openwrt-sdk-ar71xx-generic_gcc-5.5.0_musl.Linux-x86_64/build_dir/target-mips_24kc_musl/getdns-1.4.2:getdns-1.4.2 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro  -Wall -Wextra -D_BSD_SOURCE -D_DEFAULT_SOURCE -I/working/openwrt-sdk-ar71xx-generic_gcc-5.5.0_musl.Linux-x86_64/staging_dir/target-mips_24kc_musl/usr/include -I/working/openwrt-sdk-ar71xx-generic_gcc-5.5.0_musl.Linux-x86_64/staging_dir/target-mips_24kc_musl/include -I/working/openwrt-sdk-ar71xx-generic_gcc-5.5.0_musl.Linux-x86_64/staging_dir/toolchain-mips_24kc_gcc-5.5.0_musl/usr/include -I/working/openwrt-sdk-ar71xx-generic_gcc-5.5.0_musl.Linux-x86_64/staging_dir/toolchain-mips_24kc_gcc-5.5.0_musl/include/fortify -I/working/openwrt-sdk-ar71xx-generic_gcc-5.5.0_musl.Linux-x86_64/staging_dir/toolchain-mips_24kc_gcc-5.5.0_musl/include  -I/working/openwrt-sdk-ar71xx-generic_gcc-5.5.0_musl.Linux-x86_64/staging_dir/target-mips_24kc_musl/usr/include  -Wpedantic -c ./gldns/keyraw.c -o keyraw.lo
./gldns/keyraw.c: In function 'gldns_key_EVP_load_gost_id':
./gldns/keyraw.c:144:6: warning: implicit declaration of function 'ENGINE_by_id' [-Wimplicit-function-declaration]
  e = ENGINE_by_id("gost");
      ^
./gldns/keyraw.c:144:4: warning: assignment makes pointer from integer without a cast [-Wint-conversion]
  e = ENGINE_by_id("gost");
    ^
./gldns/keyraw.c:147:3: warning: implicit declaration of function 'ENGINE_load_builtin_engines' [-Wimplicit-function-declaration]
   ENGINE_load_builtin_engines();
   ^
./gldns/keyraw.c:148:3: warning: implicit declaration of function 'ENGINE_load_dynamic' [-Wimplicit-function-declaration]
   ENGINE_load_dynamic();
   ^
./gldns/keyraw.c:149:5: warning: assignment makes pointer from integer without a cast [-Wint-conversion]
   e = ENGINE_by_id("gost");
     ^
./gldns/keyraw.c:155:6: warning: implicit declaration of function 'ENGINE_set_default' [-Wimplicit-function-declaration]
  if(!ENGINE_set_default(e, ENGINE_METHOD_ALL)) {
      ^
./gldns/keyraw.c:155:28: error: 'ENGINE_METHOD_ALL' undeclared (first use in this function)
  if(!ENGINE_set_default(e, ENGINE_METHOD_ALL)) {
                            ^
./gldns/keyraw.c:155:28: note: each undeclared identifier is reported only once for each function it appears in
./gldns/keyraw.c:156:3: warning: implicit declaration of function 'ENGINE_finish' [-Wimplicit-function-declaration]
   ENGINE_finish(e);
   ^
./gldns/keyraw.c:157:3: warning: implicit declaration of function 'ENGINE_free' [-Wimplicit-function-declaration]
   ENGINE_free(e);
   ^
make[4]: *** [keyraw.lo] Error 1

**Comment by hanvinke** [comment]
Since none of the devs reacted, might this have to do with OpenSSL 1.1.0 and later no longer include the GOST engine?

**Comment by wtoorop** [comment]
I suspect that this issue is due to a OpenSSL compiled with a limited set of functionality (I vaguely remember seeing something similar before (also with OpenWRT?)).
To reproduce I need a similarly limited OpenSSL install (which I do not have readily available yet :( )

**Comment by hanvinke** [comment]
@wtoorop 
Thanks for the reaction, could it also be possible that the gost version used is just incompatible, since it had to be manually added? Maybe building again helps? https://github.com/gost-engine/engine/wiki/Building

**Comment by wtoorop** [comment]
I don't think so no.  OpenSSL engines are used for more than just gost, and the error is about missing function primitives and constants for enabling OpenSSL engines all together.
This looks more like pr #387 ... in fact that PR also talks about OpenWRT and OpenSSL compiled with the no-engines flag...

**Comment by wtoorop** [comment]
I really tried very hard, but cannot reproduce.  At least not with openssl-1.0.2o configured with the `no-engines` and `no-deprecated` options.  I suppose I should try to build an OpenWRT to be able to reproduce...

**Comment by wtoorop** [comment]
@iamperson347 Does this patch help?
  - [0001-Wild-guess-at-OpenSSL-without-engine-support.patch.gz](https://github.com/getdnsapi/getdns/files/2094577/0001-Wild-guess-at-OpenSSL-without-engine-support.patch.gz)

**Comment by iamperson347** [comment]
Not sure if this helps.... This is currently the proposed configuration for openwrt openssl1.1.*

https://github.com/cotequeiroz/openwrt-1/tree/803f8beaef92fbc419e690451c9a28c654e50f46/package/libs/openssl

The makefile is specific to openwrt build tools, but this might give some ideas on how it is being built.

**Comment by iamperson347** [comment]
@wtoorop I'll try that patch when I get a chance later today.

**Comment by iamperson347** [comment]
Hi @wtoorop, sorry it took so long for me to get to testing this. This did allow getdns to compile successfully. I still need to attempt to test out the resulting libs though.

Would this patch break anything when compiled against 1.0 openssl vs 1.1? Trying to determine if it makes sense to submit this patch to openwrt package repo or hold off... Openssl 1.1.0h still hasn't been merged yet since they are still trying to check all the packages for compatibility, etc.

**Comment by neheb** [comment]
Small update: OpenWrt will soon be switching to 1.1.1 by default. This is what I currently use to get getdns to compile: https://github.com/neheb/packages/commit/795659640d46d671b1eecec5ce45bf83124cd762

@wtoorop it works.
